k2/csrc/connect.cu

```diff
@@ -0,0 +1,519 @@
+/**
+ * Copyright      2021  Xiaomi Corporation (authors: Wei Kang)
+ *
+ * See LICENSE for clarification regarding multiple authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <limits>
+#include <vector>
+
+#include "k2/csrc/array_ops.h"
+#include "k2/csrc/context.h"
+#include "k2/csrc/device_guard.h"
+#include "k2/csrc/fsa_algo.h"
+#include "k2/csrc/fsa_utils.h"
+#include "k2/csrc/thread_pool.h"
+
+namespace k2 {
+
+class Connecter {
+ public:
+  /**
+     Connector object.  You should call Connect() after
+     constructing it.  Please see Connect() declaration in header for
+     high-level overview of the algorithm.
+
+       @param [in] fsas    A vector of FSAs; must have 3 axes.
+   */
+  explicit Connecter(FsaVec &fsas) : c_(fsas.Context()), fsas_(fsas) {
+    K2_CHECK_EQ(fsas_.NumAxes(), 3);
+    int32_t num_states = fsas_.shape.TotSize(1);
+    accessible_ = Array1<char>(c_, num_states, 0);
+    coaccessible_ = Array1<char>(c_, num_states, 0);
+  }
+
+  /*
+    Computes the next batch of states
+         @param [in] cur_states  Ragged array with 2 axes, containing
+    state-indexes (idx01) into fsas_.
+         @return   Returns the states which, after processing.
+   */
+  std::unique_ptr<Ragged<int32_t>> GetNextBatch(Ragged<int32_t> &cur_states) {
+    NVTX_RANGE(K2_FUNC);
+    // Process arcs leaving all states in `cur`
+
+    // First figure out how many arcs leave each state.
+    // And set accessible for each state
+    Array1<int32_t> num_arcs_per_state(c_, cur_states.NumElements() + 1);
+    int32_t *num_arcs_per_state_data = num_arcs_per_state.Data();
+    const int32_t *fsas_row_splits2_data = fsas_.RowSplits(2).Data(),
+                  *states_data = cur_states.values.Data();
+    char *accessible_data = accessible_.Data();
+    K2_EVAL(
+        c_, cur_states.NumElements(), lambda_set_arcs_and_accessible_per_state,
+        (int32_t states_idx01)->void {
+          int32_t fsas_idx01 = states_data[states_idx01],
```

Comment:
IIRC, for a Ragged array with index `[fsa][state][arc]`, according to our naming convention, we usally call this as `state_idx01` instead of `fsa_idx01`. So it's better if we can follow the naming convention, e.g. fsa_idx0, state_idx01, arc_idx012, that is, `theThingWeIndex_idx0[123..]`.  Of course you need find another name for `states_idx01` here (`idx_01` may be fine here).

Please also check other names below to follow the naming convention.

Reply:
I follow the index schema according to `util.h`,  and didn't find any convention relates to the prefix.  I thought all the indexes are with the same prefix (say `fsa`),  for `[fsa][state][arc]`  `fsa_idx01` is the global index of a state and `fsa_idx1` is the local index of a state.

Will check the naming convention in the other source files and follow.